scaleway/resource_token.go

```diff
@@ -0,0 +1,133 @@
+package scaleway
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+	api "github.com/nicolai86/scaleway-sdk"
+)
+
+func resourceScalewayToken() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceScalewayTokenCreate,
+		Read:   resourceScalewayTokenRead,
+		Update: resourceScalewayTokenUpdate,
+		Delete: resourceScalewayTokenDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"email": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				Description: "The account email. Defaults to registered user.",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				Description: "The token description.",
+			},
+			"password": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "User password, in case a login is require",
+			},
+			"expires": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "Defines if the token is set to expire",
+			},
+			"expiration_date": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The tokens expiration date",
+			},
+		},
+	}
+}
+
+func resourceScalewayTokenCreate(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+
+	mu.Lock()
+	email := ""
+	if mail, ok := d.GetOk("email"); ok {
+		email = mail.(string)
+	} else {
+		user, err := scaleway.GetUser()
+		if err != nil {
+			return err
+		}
+		email = user.Email
+	}
+
+	token, err := scaleway.CreateToken(&api.CreateTokenRequest{
+		Email:    email,
+		Password: d.Get("password").(string),
+		Expires:  d.Get("expires").(bool),
+	})
+	mu.Unlock()
+	if err != nil {
+		return err
+	}
+
+	d.SetId(token.ID)
+	return resourceScalewayTokenUpdate(d, m)
+}
+
+func resourceScalewayTokenRead(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+
+	token, err := scaleway.GetToken(d.Id())
+	if err != nil {
+		if serr, ok := err.(api.APIError); ok {
+			if serr.StatusCode == 404 {
+				d.SetId("")
+				return nil
+			}
+		}
+		return err
+	}
+	d.Set("description", token.Description)
+	d.Set("expiration_date", token.Expires)
+
+	return nil
+}
+
+func resourceScalewayTokenUpdate(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+
+	mu.Lock()
+	defer mu.Unlock()
+
+	if d.HasChange("description") || d.HasChange("expires") {
+		token, err := scaleway.UpdateToken(&api.UpdateTokenRequest{
+			ID:          d.Id(),
+			Expires:     d.Get("expires").(bool),
+			Description: d.Get("description").(string),
+		})
+		if err != nil {
+			return err
+		}
+		d.Set("description", token.Description)
+		d.Set("expiration_date", token.Expires)
```

Comment:
we shouldn't need these here since we're calling the Read function anyway - so can we remove them?

Reply:
good point!